### PR TITLE
adds more sanity checks to shreds

### PR DESCRIPTION
**Pull request by behzadnouri**
#### Problem
shreds inner fields need sanity checks.

#### Summary of Changes
added sanity checks

### CODE REVIEW

File: ledger/src/shred.rs [hunk 1]
```diff
@@ -448,33 +442,68 @@ impl Shred {
...
 
     pub fn index(&self) -> u32 {
         self.common_header.index
     }
 
+    pub(crate) fn fec_set_index(&self) -> u32 {
+        self.common_header.fec_set_index
+    }
+
+    // Returns true if the shred sanity checks.
```
Comment (carllin) [inline ledger/src/shred.rs:469]: nit: `if the shred passes sanity checks`

File: ledger/src/shred.rs [hunk 2]
```diff
@@ -448,33 +442,68 @@ impl Shred {
         self.common_header.slot
     }
 
-    pub fn parent(&self) -> Option<Slot> {
+    pub fn parent(&self) -> Result<Slot> {
         match self.shred_type() {
             ShredType::Data => {
-                let parent_offset = Slot::try_from(self.data_header.parent_offset);
-                self.slot().checked_sub(parent_offset.ok()?)
+                // TODO What if parent_offset is zero and slot != 0?
```
Comment (carllin) [inline ledger/src/shred.rs:448]: I think the call to this filter in window service will drop the shred https://github.com/solana-labs/solana/blob/0fd8e3c5d749832612f0e44c624f05a30e1ab48b/core/src/window_service.rs#L167

```
if parent_slot >= slot {
   return false;
}
```

**Review by carllin — DISMISSED** [review]

File: ledger/src/shred.rs (hunk 2 above)
Comment (behzadnouri) [reply ledger/src/shred.rs]: done

File: ledger/src/shred.rs (hunk 1 above)
Comment (behzadnouri) [reply ledger/src/shred.rs]: done!

**Pull request merged**